diff -crB a/drivers/net/wireless/wl12xx/wl12xx.h b/drivers/net/wireless/wl12xx/wl12xx.h
*** a/drivers/net/wireless/wl12xx/wl12xx.h	2011-12-04 09:19:23.702269996 +0200
--- b/drivers/net/wireless/wl12xx/wl12xx.h	2011-12-04 09:18:45.082258053 +0200
***************
*** 133,149 ****
  
  
  
! #define WL1271_FW_NAME "ti-connectivity/wl1271-fw-multirole-roc.bin"
! #define WL1271_PLT_FW_NAME "ti-connectivity/wl1271-fw-multirole-plt.bin"
! #define WL128X_FW_NAME "ti-connectivity/wl128x-fw-multirole-roc.bin"
! #define WL128X_PLT_FW_NAME "ti-connectivity/wl128x-fw-multirole-plt.bin"
  
  /*
   * wl127x and wl128x are using the same NVS file name. However, the
   * ini parameters between them are different.  The driver validates
   * the correct NVS size in wl1271_boot_upload_nvs().
   */
! #define WL12XX_NVS_NAME "ti-connectivity/wl1271-nvs.bin"
  
  #define WL1271_TX_SECURITY_LO16(s) ((u16)((s) & 0xffff))
  #define WL1271_TX_SECURITY_HI32(s) ((u32)(((s) >> 16) & 0xffffffff))
--- 133,149 ----
  
  
  
! #define WL1271_FW_NAME "wl1271-fw-multirole-roc.bin"
! #define WL1271_PLT_FW_NAME "wl1271-fw-multirole-plt.bin"
! #define WL128X_FW_NAME "wl128x-fw-multirole-roc.bin"
! #define WL128X_PLT_FW_NAME "wl128x-fw-multirole-plt.bin"
  
  /*
   * wl127x and wl128x are using the same NVS file name. However, the
   * ini parameters between them are different.  The driver validates
   * the correct NVS size in wl1271_boot_upload_nvs().
   */
! #define WL12XX_NVS_NAME "wl1271-nvs.bin"
  
  #define WL1271_TX_SECURITY_LO16(s) ((u16)((s) & 0xffff))
  #define WL1271_TX_SECURITY_HI32(s) ((u32)(((s) >> 16) & 0xffffffff))
