diff -ur compat-wireless/drivers/net/wireless/wl12xx/sdio.c compat-wireless.semc/drivers/net/wireless/wl12xx/sdio.c
--- compat-wireless/drivers/net/wireless/wl12xx/sdio.c	2011-12-22 07:16:11.744036003 +0200
+++ compat-wireless.semc/drivers/net/wireless/wl12xx/sdio.c	2011-12-22 07:14:57.794034922 +0200
@@ -259,8 +259,8 @@
 	/* Grab access to FN0 for ELP reg. */
 	func->card->quirks |= MMC_QUIRK_LENIENT_FN0;
 
-	/* Use block mode for transferring over one block size of data */
-	func->card->quirks |= MMC_QUIRK_BLKSZ_FOR_BYTE_MODE;
+//	/* Use block mode for transferring over one block size of data */
+//	func->card->quirks |= MMC_QUIRK_BLKSZ_FOR_BYTE_MODE;
 
 	wlan_data = wl12xx_get_platform_data();
 	if (IS_ERR(wlan_data)) {
