diff -ur compat-wireless/drivers/net/wireless/wl12xx/sdio.c compat-wireless-wl12xx/drivers/net/wireless/wl12xx/sdio.c
--- compat-wireless/drivers/net/wireless/wl12xx/sdio.c	2012-01-27 15:02:37.000000000 +0200
+++ compat-wireless-wl12xx/drivers/net/wireless/wl12xx/sdio.c	2011-12-20 18:29:47.000000000 +0200
@@ -172,7 +172,7 @@
 	/* If enabled, tell runtime PM not to power off the card */
 	if (pm_runtime_enabled(&func->dev)) {
 		ret = pm_runtime_get_sync(&func->dev);
-		if (ret)
+		if (ret < 0)
 			goto out;
 	} else {
 		/* Runtime PM is disabled: power up the card manually */
@@ -252,8 +252,8 @@
 	/* Grab access to FN0 for ELP reg. */
 	func->card->quirks |= MMC_QUIRK_LENIENT_FN0;
 
-//	/* Use block mode for transferring over one block size of data */
-//	func->card->quirks |= MMC_QUIRK_BLKSZ_FOR_BYTE_MODE;
+	/* Use block mode for transferring over one block size of data */
+	//func->card->quirks |= MMC_QUIRK_BLKSZ_FOR_BYTE_MODE;
 
 	wlan_data = wl12xx_get_platform_data();
 	if (IS_ERR(wlan_data)) {
@@ -294,6 +294,8 @@
 	}
 	disable_irq(wl->irq);
 
+	sdio_set_host_pm_flags(func, MMC_PM_KEEP_POWER);
+
 	ret = wl1271_init_ieee80211(wl);
 	if (ret)
 		goto out_irq;
@@ -306,6 +308,7 @@
 
 	/* Tell PM core that we don't need the card to be powered now */
 	pm_runtime_put_noidle(&func->dev);
+	mmc_power_save_host(func->card->host);
 
 	return 0;
 
@@ -321,9 +324,13 @@
 static void __devexit wl1271_remove(struct sdio_func *func)
 {
 	struct wl1271 *wl = sdio_get_drvdata(func);
+	struct mmc_host *host = func->card->host;
 
 	/* Undo decrement done above in wl1271_probe */
 	pm_runtime_get_noresume(&func->dev);
+	mmc_power_restore_host(func->card->host);
+
+	host->pm_flags = host->pm_flags & ~MMC_PM_KEEP_POWER;
 
 	wl1271_unregister_hw(wl);
 	if (wl->irq_wake_enabled) {
@@ -357,6 +364,7 @@
 			goto out;
 		}
 
+		printk("\n\nSetting MMC_PM_KEEP_POWER\n");
 		/* keep power while host suspended */
 		ret = sdio_set_host_pm_flags(func, MMC_PM_KEEP_POWER);
 		if (ret) {
@@ -376,6 +384,7 @@
 	struct sdio_func *func = dev_to_sdio_func(dev);
 	struct wl1271 *wl = sdio_get_drvdata(func);
 
+	printk("%s\n", __func__);
 	wl1271_debug(DEBUG_MAC80211, "wl1271 resume");
 	if (wl->wow_enabled) {
 		/* claim back host */
